### ISSUE TITLE
FileNotFoundException on ZTE Grand X Max 2

### ISSUE THREAD

**Issue by oprisnik** [facebook/SoLoader#15]
Fresco is using SoLoader and we got a bug report (https://github.com/facebook/fresco/issues/2176) that on the ZTE Grand X Max 2 (API 23), SoLoader cannot be initialized:

```
Caused by java.io.FileNotFoundException: /data/user/0/com.remind101/lib-main/dso_lock: open failed: EACCES (Permission denied)
at libcore.io.IoBridge.open(IoBridge.java:452)
at java.io.FileOutputStream.(FileOutputStream.java:87)
at java.io.FileOutputStream.(FileOutputStream.java:72)
at com.facebook.soloader.FileLocker.(FileLocker.java:28)
at com.facebook.soloader.FileLocker.lock(FileLocker.java:24)
at com.facebook.soloader.UnpackingSoSource.prepare(UnpackingSoSource.java:393)
at com.facebook.soloader.SoLoader.initImpl(SoLoader.java:211)
at com.facebook.soloader.SoLoader.init(SoLoader.java:120)
at com.facebook.soloader.SoLoader.init(SoLoader.java:104)
at com.facebook.drawee.backends.pipeline.Fresco.initialize(Fresco.java:63)
at com.facebook.drawee.backends.pipeline.Fresco.initialize(Fresco.java:46)
Caused by android.system.ErrnoException: open failed: EACCES (Permission denied)
at libcore.io.Posix.open(Posix.java)
at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
at libcore.io.IoBridge.open(IoBridge.java:438)
at java.io.FileOutputStream.(FileOutputStream.java:87)
at java.io.FileOutputStream.(FileOutputStream.java:72)
at com.facebook.soloader.FileLocker.(FileLocker.java:28)
at com.facebook.soloader.FileLocker.lock(FileLocker.java:24)
at com.facebook.soloader.UnpackingSoSource.prepare(UnpackingSoSource.java:393)
at com.facebook.soloader.SoLoader.initImpl(SoLoader.java:211)
at com.facebook.soloader.SoLoader.init(SoLoader.java:120)
at com.facebook.soloader.SoLoader.init(SoLoader.java:104)
at com.facebook.drawee.backends.pipeline.Fresco.initialize(Fresco.java:63)
at com.facebook.drawee.backends.pipeline.Fresco.initialize(Fresco.java:46)
```

It looks like ZTE has modified Android for whatever reason that prevents us from loading native libraries through SoLoader. The line where this is failing is here: https://github.com/facebook/SoLoader/blob/master/java/com/facebook/soloader/FileLocker.java#L35

**Comment by andreibalmos** [comment]
Hey guys,
what's the status with this bug, has it been solved?
I'm getting the same crash on OnePlus 5T with Android 8.0.

Can't post the stacktrace here but it's basically caused by this:
`java.io.IOException: cannot mkdir: /data/user/0/_app-package_/lib-main
at com.facebook.soloader.SoLoader.init(SourceFile:181)`

**Comment by zhaoshe** [comment]
Anyone solved this? I got the same bug when initiating SoLoader.

`Caused by: java.io.IOException: cannot mkdir: /data/user/0/app.buzz.share/lib-main
at com.facebook.soloader.k.b(k.java:146)
at com.facebook.soloader.m.a(m.java:415)
at com.facebook.soloader.SoLoader.b(SoLoader.java:266)
at com.facebook.soloader.SoLoader.a(SoLoader.java:170)
at com.facebook.soloader.SoLoader.a(SoLoader.java:153)`

**Comment by bbssyyuui** [comment]
I got this bug too, has anyone solved this?

**Comment by ouabing** [comment]
Same issue on OnePlus 8T with Android 11.
```
Caused by: java.io.IOException: cannot mkdir: /data/user/0/me.tangke.packagename/lib-main
at com.facebook.soloader.SysUtil.mkdirOrThrow(SysUtil.java:237)
at com.facebook.soloader.UnpackingSoSource.prepare(UnpackingSoSource.java:409)
at com.facebook.soloader.SoLoader.initSoSources(SoLoader.java:322)
at com.facebook.soloader.SoLoader.init(SoLoader.java:200)
at com.facebook.soloader.SoLoader.init(SoLoader.java:182)
at com.facebook.soloader.SoLoader.init(SoLoader.java:212)
```

**Comment by keletappi** [comment]
It looks like sometimes `lib-main` gets created under application's home folder with no write bit set:
```
XQ-AS52:/ $ run-as my.app.package

XQ-AS52:/data/user/0/my.app.package $ ls -alF
total 97
drwx------   8 u0_a550 u0_a550         3488 2021-09-22 17:07 ./
drwxrwx--x 631 system  system        110592 2021-09-22 17:07 ../
drwxrws--x   2 u0_a550 u0_a550_cache   3488 2021-09-22 17:07 cache/
drwxrwx--x   2 u0_a550 u0_a550         3488 2021-09-22 17:07 databases/
drwxrwx--x   4 u0_a550 u0_a550         3488 2021-09-22 17:07 files/
dr-x------   2 u0_a550 u0_a550         3488 1970-01-19 22:46 lib-main/
drwxrwx--x   2 u0_a550 u0_a550         3488 2021-09-22 17:07 no_backup/
drwxrwx--x   2 u0_a550 u0_a550         3488 2021-09-22 17:07 shared_prefs/

XQ-AS52:/data/user/0/my.app.package/lib-main $ ls -alF
total 6
dr-x------ 2 u0_a550 u0_a550 3488 1970-01-19 22:46 ./
drwx------ 8 u0_a550 u0_a550 3488 2021-09-22 17:07 ../
```
which causes SoLoader to fail when it tries to create the lock file `dso_lock` under that folder. And the folder happens to be empty. It's also bit suspicious the `lib-main` has been created with 1970 timestamp.

**Comment by simpleton** [comment]
I think this issue was fixed in v0.10.3. I will close it due to it's too old. Feel free to reopen the issue if you're still experiencing a similar issue on v0.10.3
